### ISSUE TITLE
Program got killed when running ginex

### ISSUE THREAD

**Issue by fucker-cracker** [SNU-ARC/Ginex#1]
I follow the same commands as you give in README.md, 
and everything went well until I run the final command, 
which is "` sudo PYTHONPATH=/home/user/.local/lib/python3.8/site-packages cgexec -g memory:8gb python3 -W ignore run_ginex.py --neigh-cache-size 6000000000 --feature-cache-size 6000000000 --sb-size 1500`".
After this command, the program just got killed.
I watched the htop and found that the memory limit(8g) plus swap space(4g) counldn't meet the capacity requirements the program supposed to be.
My build environment:

cpu_info: Intel(R) Xeon(R) Gold 5320 CPU  @ 2.20GHz
gpu_info: NVIDIA Tesla V100 32GB PCIe
memory: Samsung DDR4 512G（8g cgroup limit used)
swapfile:4g used
cuda: "11.6"
python:3.8.10
pytorch:1.13
nvidia-driver:510.108.03
Ubuntu:20.04.5

By the way, 
- The baseline program runs well, and got the same result (about 250s per epoch) as you present,
- To make the ginex program run, I quit the memory limit, which means I run the ginex in a enough mem env.But the result seems different from yours(The desired effect was not achieved.And the max mem use up to 40g).

**Comment by ilil96** [comment]
Hi,

Could you please share the console output after running Ginex with --verbose flag?
I guess I need to check in which part of the program the process gets killed.

**Comment by fucker-cracker** [comment]
Thanks for replying!
I just ran the program with --verbose flag,
`sudo PYTHONPATH=/home/user/.local/lib/python3.8/site-packages cgexec -g memory:8gb python3 -W ignore run_ginex.py --neigh-cache-size 6000000000 --feature-cache-size 6000000000 --sb-size 1500 --verbose`
And the console output is:
`Preparing dataset...`
`Killed`

**Comment by ilil96** [comment]
Can you share 'conf.json' file in the dataset directory?

**Comment by fucker-cracker** [comment]
> Can you share 'conf.json' file in the dataset directory?

OK,the conf.json file path is "./Ginex/dataset/ogbn-papers100M-ginex/conf.json"
and the content is:
`{"num_nodes": 111059956, "indptr_shape": [111059957], "indptr_dtype": "int64", "indices_shape": [1615685872], "indices_dtype": "int64", "features_shape": [111059956, 128], "features_dtype": "float32", "labels_shape": [111059956, 1], "labels_dtype": "float32", "num_classes": 172}`

**Comment by ilil96** [comment]
It seems like the memory leak occurs when loading dataset and that's weird as baseline and Ginex actually use the same data loading code.
Could you check which line among line 56, line 57, and line 58 causes the error in run_ginex.py?
You may execute code line by line by using an interactive debugger like pdb or you may just insert print command between lines.

**Comment by fucker-cracker** [comment]
> It seems like the memory leak occurs when loading dataset and that's weird as baseline and Ginex actually use the same data loading code. Could you check which line among line 56, line 57, and line 58 causes the error in run_ginex.py? You may execute code line by line by using an interactive debugger like pdb or you may just insert print command between lines.

OK, I just simplely add "print(" x line start!")  before line 56,57,58.
`print("56 line start!")`
`mmapped_features = dataset.get_mmapped_features()`
`print("57 line start!")`
`indptr, indices = dataset.get_adj_mat()`
`print("58 line start!")`
`labels = dataset.get_labels()`
And with the same running command, the output is:
Preparing dataset...
56 line start!
57 line start!
Killed
So it seems like line 57 causes the memory leak

**Comment by fucker-cracker** [comment]
My bad
I forgot that I added some code due to this warning when running `python3 create_neigh_cache.py --neigh-cache-size 6000000000`
`UserWarning: The given NumPy array is not writable, and PyTorch does not support non-writable tensors. This means writing to this tensor will result in undefined behavior. You may want to copy the array to protect its data or make it writable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:199.)
  indices = torch.from_numpy(indices)`
I changed `indices = torch.from_numpy(indices)` to `torch.from_numpy(indices.copy())`
Now I quit this change but it still got killed, and this is the new console output:
`Preparing dataset...`
`Done!`

`==============================`
`Running Epoch 0...`
`Running 0th superbatch of total 1 superbatches  `                                                                                                         
`Step 1: Superbatch Sample   `                                                                                                                             
`Epoch 00:   0%|                                                                                                              | 0/1207179 [00:00<?, ?it/s]`
`Killed`

**Comment by ilil96** [comment]
Have you made any other modification to create_neigh_cache.py?
Please share 'nctbl_size_6000000000_conf.json' and 'nc_size_6000000000_conf.json' as well as the size of 'nctbl_size_6000000000.dat' and 'nc_size_6000000000.dat' in the dataset directory.

**Comment by fucker-cracker** [comment]
> Have you made any other modification to create_neigh_cache.py? Please share 'nctbl_size_6000000000_conf.json' and 'nc_size_6000000000_conf.json' as well as the size of 'nctbl_size_6000000000.dat' and 'nc_size_6000000000.dat' in the dataset directory.

Sorry again for that mistake. To prevent other modifications I may make, I removed the last working directory and git clone the repo agin.
Here is the info:
content of nctbl_size_6000000000_conf.json:
`{"shape": [111059956], "dtype": "int64"}`
content of nc_size_6000000000_conf.json:
`{"shape": [638940044], "dtype": "int64"}`
size of nctbl_size_6000000000.dat： 848M
size of nc_size_6000000000.dat: 4.8G

**Comment by ilil96** [comment]
Has any trace file been generated in trace directory?

**Comment by fucker-cracker** [comment]
> Has any trace file been generated in trace directory?

Yes, after running ginex program, a trace directory been generated
`└── trace`
`    └── 2023_04_29_22_39_33`
It seems no file in it

**Comment by ilil96** [comment]
Is it empty? If not, how many files are there in the directory and how big are they?

**Comment by fucker-cracker** [comment]
> Is it empty? If not, how many files are there in the directory and how big are they?

It is empty, no files in it

**Comment by ilil96** [comment]
That's strange...
Can you confirm that the program is killed by OOM also for this time?
If so, line 107 or line 108 in run_ginex.py may be the one that causes the problem.
Can you check which one is the problem?

**Comment by fucker-cracker** [comment]
> That's strange... Can you confirm that the program is killed by OOM also for this time? If so, line 107 or line 108 in run_ginex.py may be the one that causes the problem. Can you check which one is the problem?

Yes,I can confirm that the program is killed by OOM , mainly for two reasons;
1, After I removed the memory limit, the program ran normally .
2, I carefully observed the HTOP monitoring interface, and when SwapSpace just hit the 4G cap, the console immediately output “killed”

I checked which line was causing the problem and found that it was neither line 107 nor line 108 of the issue，and it turn out 
program got killed during this loop:
`for step, _ in enumerate(loader):`
 `       if i != 0 and step == 0:`
 `          cache.pass_3(iterptr, iters, initial_cache_indices)`

**Comment by ilil96** [comment]
Have you tried upgrading your pytorch version?

**Comment by fucker-cracker** [comment]
> Have you tried upgrading your pytorch version?

My current pytorch version v1.13.1 is already the most recent version under major version "1",  you mean I upgrade it to v2.0 or  change it to a lower version

**Comment by ilil96** [comment]
Oh I meant changing it to the lower version (1.9) on which I tested the code.
Before that, could you change the prefetch_factor argument for dataloader in line 116 in run_ginex.py to some small value like 2 and see if anything changes?
Specifically, can you check if any file is generated in the trace directory for this time with small prefetch_factor?

**Comment by fucker-cracker** [comment]
> Oh I meant changing it to the lower version (1.9) on which I tested the code. Before that, could you change the prefetch_factor argument for dataloader in line 116 in run_ginex.py to some small value like 2 and see if anything changes? Specifically, can you check if any file is generated in the trace directory for this time with small prefetch_factor?

Ok,
I changed the prefetch_factor from "1<<20" to "2", and still got the same problem.
In the trace dir, only a new dir called "trace/2023_05_01_16_15_17"  appeared , with no files in it.
I'll try the lower pytroch version later, and let you know when new results come out.
Thanks for the constant response
